### PR TITLE
[SPARK-16391][SQL] KeyValueGroupedDataset.reduceGroups should support partial aggregation

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

`KeyValueGroupedDataset.reduceGroups` is currently implemented via `flatMapGroups`, which does not support partial aggregation and so is very inefficient.

`KeyValueGroupedDataset.reduceGroups` should support partial aggregation. This PR implements it with `Aggregator`.
## How was this patch tested?

Existing tests.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala [hunk 1]
```diff
@@ -177,10 +178,33 @@ class KeyValueGroupedDataset[K, V] private[sql](
...
-    val func = (key: K, it: Iterator[V]) => Iterator((key, it.reduce(f)))
+    val encoder = encoderFor[V]
+    val intEncoder: ExpressionEncoder[Int] = ExpressionEncoder()
+    val aggregator: TypedColumn[V, V] = new Aggregator[V, (Int, V), V] {
+      def bufferEncoder: Encoder[(Int, V)] = ExpressionEncoder.tuple(intEncoder, encoder)
+      def outputEncoder: Encoder[V] = encoder
 
-    implicit val resultEncoder = ExpressionEncoder.tuple(kExprEnc, vExprEnc)
-    flatMapGroups(func)
+      def zero: (Int, V) = (0, null.asInstanceOf[V])
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: One problem with `Aggregator` here is the zero value. This PR uses an Int (can be Boolean too) to indicate if the buffer is initialized.

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: pull this out to be a generic ReduceAggregator and add unit test for it?

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: ok. I will update this.

**Comment by koertkuipers** [comment]
there is a usefulness to this `ReduceAggregator` beyond `.reduceGroups`. basically you can take any Aggregator without a zero and turn it into a valid Aggregator, with the caveat being that the result is nullable and will be null if no inputs are provided to the Aggregator.

**Comment by viirya** [comment]
ping @rxin The change is ok for you? Thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/expressions/ReduceAggregatorSuite.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.expressions
+
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder
+import org.apache.spark.sql.expressions.ReduceAggregator
+
+class ReduceAggregatorSuite extends SparkFunSuite {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/expressions/ReduceAggregatorSuite.scala]: just put this in DatasetAggregatorSuite

**Comment by rxin** [comment]
@viirya I'm going to take over the PR and play with the API a little bit.

**Comment by viirya** [comment]
Ok.

File: sql/core/src/test/scala/org/apache/spark/sql/expressions/ReduceAggregatorSuite.scala (hunk 2 above)
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/expressions/ReduceAggregatorSuite.scala]: Do I need to update this as you want to take over it?

Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/expressions/ReduceAggregatorSuite.scala]: I updated it.

**Comment by viirya** [comment]
@rxin Any thing I need to update for this? Thanks.

**Comment by viirya** [comment]
ping @rxin

**Comment by viirya** [comment]
ping @rxin any thoughts on this? It is waiting for a while. Thanks.

**Comment by rxin** [comment]
I've created a pr here for discussion, based on my playing with the API: https://github.com/apache/spark/pull/14576

**Comment by viirya** [comment]
Close this now since the pr #14576 is merged.

**Pull request closed without merging**